### PR TITLE
[MRG] inits for sr(F)GW + srGW barycenter + exemples

### PR DESCRIPTION
**Pull request by cedricvincentcuaz**
## Types of changes

- add initialization heuristics for sr(F)GW problems `semirelaxed_init_plan` in `ot.gromov._utils.py`.
- Integrate these initializations as options in sr(F)GW matching solvers by allowing `G0` to also be a string among admissible init methods.
- add `semirelaxed_gromov_barycenters` following API of `gromov_barycenters`, while adding `G0 : str` to leverage init methods in `semirelaxed_init_plan` to both init transport plans and the barycenter.
- Integrate these inits in srFGW barycenter solver too.
- Update the lp_solver of `semirelaxed_cg` in `ot.optim` : relaxing the identification of the minimum in the gradient with a small margin to take into account precision errors. This was causing issues of consistency across backends.
- remove some useless kwargs in entropic sr(F)GW solvers, which were absorbing `tol_rel` and `tol_abs` parameters used in the solver `ot.solve_gromov`, whereas only parameter `tol` is used in entropic sr(F)GW solvers.
- add exemple for both srGW and srFGW barycenters for dictionary learning.

## PR checklist

- [x] I have read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

File: ot/gromov/_utils.py [hunk 1]
```diff
@@ -436,14 +440,15 @@ def semirelaxed_init_plan(C1, C2, p, M=None, alpha=1.,
     if nx is None:
         arr = [C1, C2, p]
         if M is not None:
-            arr += arr
+            arr.append(M)
+
         nx = get_backend(*arr)
```
Comment (rflamary) [inline ot/gromov/_utils.py:445]: get_backedn aeccpts None now no need for this

File: ot/gromov/_semirelaxed.py [hunk 2]
```diff
@@ -1212,42 +1524,131 @@ def semirelaxed_fgw_barycenters(
...
         if init_X is None:
-            X = nx.zeros((N, d), type_as=ps[0])
+            stacked_features = nx.concatenate(Ys, axis=0)
+            if sklearn_import:
+                stacked_features = nx.to_numpy(stacked_features)
+                km = KMeans(n_clusters=N, random_state=random_state,
+                            n_init=1).fit(stacked_features)
+                init_X = nx.from_numpy(km.cluster_centers_)
+            else:
+                warnings.warn(
```
Comment (rflamary) [inline ot/gromov/_semirelaxed.py:1597]: raise error: install sklearn for kmeans (redo a better texte) and tests that teh error is raised

File: ot/gromov/_semirelaxed.py [hunk 3]
```diff
@@ -1212,42 +1524,131 @@ def semirelaxed_fgw_barycenters(
...
+                    "Kmeans clustering cannot be performed to init barycenter "
+                    "features, so we only randomly sample input features. "
+                    "Consider installing scikit-learn.",
+                    stacklevel=2
+                )
+                np.random.seed(random_state)
```
Comment (rflamary) [inline ot/gromov/_semirelaxed.py:1603]: just raise dont' do anything if not intsalled

File: ot/gromov/_semirelaxed.py [hunk 4]
```diff
@@ -1212,42 +1524,131 @@ def semirelaxed_fgw_barycenters(
...
+                selected_idx = np.random.choice(
+                    np.arange(stacked_features.shape[0]), size=N, replace=False)
+                init_X = stacked_features[selected_idx]
```
Comment (rflamary) [inline ot/gromov/_semirelaxed.py:1606]: test this line

File: ot/gromov/_utils.py [hunk 5]
```diff
@@ -370,6 +388,197 @@ def h2(b):
...
+    elif method == 'random':
+        np.random.seed(random_state)
+        U = np.random.uniform(0, n * m, size=(n, m))
+        U = (p / U.sum(1))[:, None] * U
+        T = nx.from_numpy(U, type_as=C1)
+
+    elif method in ['fluid', 'fluid_soft']:
+        # compute fluid partitioning on the biggest graph
+        if C_to_partition is None:
+            T, q = get_transport_from_partition(None)
```
Comment (rflamary) [inline ot/gromov/_utils.py:528]: not tested

File: ot/gromov/_utils.py [hunk 6]
```diff
@@ -370,6 +388,197 @@ def h2(b):
...
+
+            T, q = get_transport_from_partition(part)
+
+        if 'soft' in method:
+            T = (T + nx.outer(p, q)) / 2.
+
+    elif method in ['spectral', 'spectral_soft']:
+        # compute spectral partitioning on the biggest graph
+        if C_to_partition is None:
+            T, q = get_transport_from_partition(None)
```
Comment (rflamary) [inline ot/gromov/_utils.py:546]: not tested

**Review by rflamary — COMMENTED** [review]
Thanks @cedricvincentcuaz , a few comments

**Comment by cedricvincentcuaz** [comment]
Thank you for your review @rflamary, I've finished taking your comments into account.

**Pull request merged**